### ISSUE TITLE
Supprt for ALB as event source?

### ISSUE THREAD

**Issue by dmateos** [Miserlou/Zappa#1715]
Just wondering if theres any insight to if this will work and if not if it can be made to in the future? :) 

As per here: 
https://docs.aws.amazon.com/elasticloadbalancing/latest/application/lambda-functions.html

This would eliminate some timeout issues we have using API Gateway.

**Comment by PeteW** [comment]
API gateway, for my usage needs, imposes a deal-breaking 30-second hard limit on request execution time. While 99.9% of the methods meet this limit I simply cannot guarantee all times will fall within that time range now and future.

For this reason I moved away from API gateway and back to dedicated EC2 instances. The ALB route not only works around this 30-second issue but it presents a potentially graceful transition to switching to lambda.

I'm familiar with the zappa framework and would be happy to assist with implementation of support for ALB deployment. It is an architectural question for the core developers of this project. Is ALB deployment something they see as "in-scope" for zappa?

**Comment by PeteW** [comment]
I'm new to zappa source code. I was under the notion that API Gateway was inextricably woven into the core. After reading/experimenting today I was pleasantly surprised how well-organized the api-gateway stuff is logically cordoned from lambda/s3/cli. 


The `use_apigateway` as an option set to `False` convinced me that while API Gateway is a major feature, the real power is in lambda management and execution.


In this PR I've delivered functionality where we leverage the already perfect zappa deployment tools and adjust the "handler" logic to detect if the caller is API Gateway or ALB and react accordingly. This code change enables zappa to work flask functions from ALB. API Gateway concepts such as "stages", "domains", "authentication" dont currently carry over. Also I didnt add the support for auto-provisioning ELBs and targeting (currently I'm convinced cloudformation doesnt yet support this sort of configuration yet, I have an open support ticket)


To get it working, you will need to:
* Set `use_apigateway` option to false (unless you want both APIGateway AND ELB which presumably should work but most people wouldnt want that)
* Run zappa deploy
* Manually establish your own ALB. Ensure the subnets align with your lambda function and the security groups are reachable
* Manually configure the targeting to route to the lambda function (I recommend doing this from EC2, not lambda so you can have human-readable names)

**Comment by PeteW** [comment]
Bump: anyone around to review/merge this?

**Comment by kennovak-zy** [comment]
I've been using this in my shop for a few weeks, and it's working well.  Super-useful!

**Comment by kennovak-zy** [comment]
I found one potential issue.  This works well for GET and POST methods, but we have a problem with OPTIONS.  (OPTIONS is used for CORS pre-flight checks, when a browser is addressing the endpoint.) We can see that the request is delivered from the browser through the ALB into the Lambda, and the Lambda runs normally.  But when the response is sent to the ALB, it marks it in its log as LambdaInvalidResponse, and it converts the Lambda's 200 response code to a 502, which is what it sends to the browser. 

I've opened a ticket at AWS about this, to get more info.  But I suspect it's a problem with the response handling in this PR, since I've seen similar behavior from other tests of the ALB with Lambda.  Thanks!

**Comment by PeteW** [comment]
Understood. Looking in to this

**Comment by PeteW** [comment]
The issue is clear enough when you reconcile the hander with the AWS ALB specifications. First look here:
https://docs.aws.amazon.com/elasticloadbalancing/latest/application/lambda-functions.html#respond-to-load-balancer
The key point is:
```
The response from your Lambda function must include the Base64 encoding status, status code, status description, and headers. You can omit the body. The statusDescription header must contain the status code and reason phrase, separated by a single space. 
```
Currently the so-called handler module supplies the header and status-code in the response, but it never supplies the status-description and the base64-encoding is conditionally supplied if and only if the response has a body.

What you have encountered is the case where ELB is silently failing on account of isBase64Encoded missing from the response because your OPTIONS response has no body. I need to make further modifications to ensure the ALB specification is honored in the response.

Now with that in mind consider my test harness. Bare bones, I hope you agree.
```
import logging
from flask import Flask, request, make_response

app = Flask(__name__)
logging.basicConfig()
logger = logging.getLogger(__name__)
logger.setLevel(logging.DEBUG)

@app.route('/', methods=['GET', 'POST', 'OPTIONS'])
def lambda_handler(event=None, context=None):
    logger.info('Lambda function invoked index()')
    if request.method == 'OPTIONS':
        logger.info('OPTIONS method detected returning CORS response')
        response = make_response()
        response.headers.add('Access-Control-Allow-Origin', '*')
        response.headers.add('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        response.headers.add('Access-Control-Allow-Methods', '*')
        return response
    else:
        return 'Flask says Hello!!'

if __name__ == '__main__':
    app.run(debug=True)
```

Consider this response AFTER the patch:
```
[pw@localhost ~]$ curl 'http://myalb/' -X OPTIONS -v
> OPTIONS / HTTP/1.1
> User-Agent: curl/7.29.0
> Host: me
> Accept: */*
>
< HTTP/1.1 200 OK
< Server: awselb/2.0
< Date: Wed, 16 Jan 2019 19:44:13 GMT
< Content-Type: text/html; charset=utf-8
< Content-Length: 0
< Connection: keep-alive
< Access-Control-Allow-Methods: *
< Access-Control-Allow-Origin: *
<
* Connection #0 to host left intact
```

**Comment by PeteW** [comment]
The [commit is #28d00e8](https://github.com/PeteW/Zappa/commit/2f3a71699446e484cbdae165c7e68090b15e18f2) which will address your issue.

Thanks for the feedback and I feel more confident this will match your needs.

**Comment by jedestep** [comment]
Thanks for your prompt help @PeteW, the patch you submitted fixed our problem. We've now tested your PR on a live environment and can confirm it's working as intended.

**Comment by jedestep** [comment]
If it helps, +1 for merging this PR and including it in the next release. This change has been a big help for us and we're currently relying on it in some environments. I'd like to return us to building off mainline Zappa instead of a fork, but we do really need this PR merged before we can do so.

**Comment by nitrocode** [comment]
@PeteW thanks for adding the change to use ALB instead of API Gateway. Any update on what the necessary steps are on how this can be merged?

**Comment by PeteW** [comment]
@nitrocode thanks for your inquiry.
The standard protocol is to request review of the PR, then one of the maintainers of the project review the code changes. Once they pass review the PR is merged back to the master branch, and, eventually a release is built.

As of this morning 2019-02-25 this pull request is in a state of waiting for code review from one of the project maintainers. It has been awaiting review for over 10 weeks now.

You have two options. One is to wait for the review/merge/release process to complete. The other is to clone this branch and use it in your virtual environment. A few folks including myself are taking the second option as a temporary solution. Forking from the master branch is a calculated short term risk.

**Comment by PeteW** [comment]
@nitrocode I have written to the slack core channel on your behalf requesting update on plans to review PR

**Comment by jneves** [comment]
Merged on #1807 - thanks a lot!

**Comment by markwatson** [comment]
I know this issue is closed, but I have a question regarding implementation. It looks like this created an internet-facing ALB by default. Is there any way to have it create an internal one? Thanks!
